conn.go

```diff
@@ -349,72 +340,60 @@ func (c *Conn) start() error {
 	// this is because our peer can tell us the max channels they support.
 	c.channels = bitmap.New(uint32(c.channelMax))
 
-	// start multiplexor and writer
-	go c.mux()
 	go c.connWriter()
+	go c.connReader()
 
 	return nil
 }
 
 // Close closes the connection.
 func (c *Conn) Close() error {
-	c.closeMuxOnce.Do(func() { close(c.closeMux) })
-	err := c.err()
+	c.closeOnce.Do(func() { c.close() })
+
 	var connErr *ConnError
-	if errors.As(err, &connErr) && connErr.RemoteErr == nil && connErr.inner == nil {
+	if errors.As(c.doneErr, &connErr) && connErr.RemoteErr == nil && connErr.inner == nil {
 		// an empty ConnectionError means the connection was closed by the caller
-		// or as requested by the peer and no error was provided in the close frame.
 		return nil
 	}
-	return err
+
+	// there was an error during shut-down or connReader/connWriter
+	// experienced a terminal error
+	return c.doneErr
 }
 
-// close should only be called by conn.mux.
+// close is called once, either from Close() or when connReader/connWriter exits
 func (c *Conn) close() {
-	close(c.done) // notify goroutines and blocked functions to exit
+	defer close(c.done)
```

Comment:
Could we move the `c.closeOnce.Do(func())` call into the `func (c *Conn) close()` function? Then you'd never have to worry about where or how often it's called.

Reply:
Good idea, fixed.